File: JavaScript/281-Zigzag Iterator.js
/**
 * @constructor
 * @param {Integer[]} v1
 * @param {Integer[]} v1
 */
var ZigzagIterator = function ZigzagIterator(v1, v2) {
    this.arr = new Array();
    for (let v of [v1, v2]){
        if (v.length > 0) this.arr.push(v);
    }
    this.queue = new Array();    
    for (let vIndex=0; vIndex < this.arr.length; vIndex++){
        this.queue.push([vIndex, 0]);
    }
};


/**
 * @this ZigzagIterator
 * @returns {boolean}
 */
ZigzagIterator.prototype.hasNext = function hasNext() {
    return this.queue.length > 0
};

/**
 * @this ZigzagIterator
 * @returns {integer}
 */
ZigzagIterator.prototype.next = function next() {
    [vIndex, itemIndex] = this.queue.shift();
    if (itemIndex < this.arr[vIndex].length-1){
        this.queue.push([vIndex, itemIndex+1]);
    }
    return this.arr[vIndex][itemIndex];
};

/**
 * Your ZigzagIterator will be called like this:
 * var i = new ZigzagIterator(v1, v2), a = [];
 * while (i.hasNext()) a.push(i.next());
*/

File: C++/281-Zigzag Iterator.cpp
class ZigzagIterator {
public:
    queue<pair<int, vector<int>&>> q;
    ZigzagIterator(vector<int>& v1, vector<int>& v2) {
        if (v1.size() > 0) q.push({0, v1});
        if (v2.size() > 0) q.push({0, v2});
    }

    int next() {
        int res = 0;
        if (!q.empty()) {
            auto [idx, v] = q.front();
            q.pop();
            res = v[idx];
            int nxtIdx = idx+1;
            if (nxtIdx < v.size()) {
                q.push({nxtIdx, v});
            }
        }
        return res;
    }

    bool hasNext() {
        return !q.empty();
    }
};

/**
 * Your ZigzagIterator object will be instantiated and called as such:
 * ZigzagIterator i(v1, v2);
 * while (i.hasNext()) cout << i.next();
 */